content/blog/2019-11-07-immix/index.md

```diff
@@ -0,0 +1,210 @@
+(1) refining this introduced mark-region collector to improve utilization of a region;
+(2) introducing a defragmentation mechanism, 
+which is presented in the next section.
+
+Immix refines mark-region by operating at two levels, 
+coarse-grained *blocks*, and fine-grained *lines*. 
+Comparing to the original mark-region collector, 
+the differences are: 
+(1) Objects may span lines; 
+(2) during reclamation,
+immix also identifies free lines in partially free blocks (formerly unavailable blocks); 
+(3) during allocation, 
+immix first tries to find contiguous free lines that are large enough for the allocates. 
+If it cannot find suitable lines, 
+then it allocates into free blocks.
+
+This paper elaborates more on the allocation process of immix, 
+which is called *steady state allocation*. 
+The allocator is divided into two parts: 
+a global allocator that maintains a pool of free blocks,
+and a thread-local allocator that maintains free lines. 
+
+![002.png](./002.png)
+
+The figure above illustrates the basic immix heap organization during steady state allocation. 
+The allocator maintains two pointers (Bump Pointer Cursor and Bump Pointer Limit) that indicate the left and right bound of the current hole (contiguous free lines),
+the right bound is either right before next living line or the end of this block, since an object cannot span blocks. 
+
+#### Implementation Details
+
+This paper discusses many details of its implementation; 
+here are some remarkable ones about mark-region.
+
+1. **Parallelism**. Immix implementations are parallel. The thread-local allocators are mostly unsynchronized while the global allocator is synchronized. Therefore, they follow a design pattern to maximize fast, unsynchronized thread-local activities and minimizes synchronized global activities. As a result, immix's mark-region collector requires hardly any synchronization.
+2. **Demand Driven Overflow Allocation**. Each immix allocator is paired with an *overflow allocator* that is also a contiguous allocator, however, it always uses empty blocks. If immix cannot allocate a *medium* object (an object with size greater than a line) into a block that contains free lines, immix allocates it with the overflow allocator.
+
+### Opportunistic Evacuation
+
+The mark-region collector, 
+like a mark-sweep one, 
+is non-moving and thus subject to fragmentation. 
+Immix uses a defragmentation mechanism called *opportunistic evacuation*.
+
+Actually, before the evacuation process, 
+methods are needed to determine when to trigger defragmentation and how to select candidate blocks. 
+This part is left to be discussed in implementation details, 
+let's suppose we know the set of candidate blocks for now. 
+
+This evacuation process is: 
+when the collector encounters a live object in a candidate block, 
+it *opportunistically* evacuates the object. 
+That is, it only evacuates an object if prior evacuation has not exhausted all target space. 
```

Comment:
Rephrased, now sure if clear enough, 
"That is, it only evacuates an object if it can successfully perform the following: 
the collector allocates this object in the same way it allocates an object created by the mutator,
except it doesn't allocate into defragmentation candidate blocks. "

Reply:
Sounds good to me!